### DISCUSSION TITLE
Using abbreviations with special characters

### DISCUSSION THREAD

**Discussion by tomvachon** [mkdocs/mkdocs#3714] [Q&A]
Hi Github Hive Mind!

I have a need to have abbreviations work with a # as part of the string to match on.  No matter what I am doing, I can't seem to get it to match.  For instance in my text I would have:
`ABC#`

The goal would be to have that have a tooltip saying `ABC1` with a tooltip.  That way whenever the number changes I can update one spot via snippets and they all get updated.  

I kind of found a workaround but it requires a ton of find and replace.  It would be: `ABC#:material-information-outline: {title="ABC 1"}`. I could have this on liek 10 pages minimum and I'd have to find and replace all of them.  I guess if I could have a way to auto-replace `ABC#` with that full string from a single "dictionary" that would also fit the bill.

Thoughts?

**Comment by pawamoy** [comment]
Sorry, it's unclear to me what you're trying to achieve. Can you show some inputs, the output after building your docs, and what you would expect instead?

**Comment by tomvachon** [comment]
in my abbrevations.md I'd like to have:

*[ABC#]: ABC1

Then whenever `ABC#` appears in a page, it would have a tooltip for `ABC1`. 

I'm using mkdocs to build a reference site for flight simulation ATC and as the FAA updates procedures they increment the number of the procedure.  That procedure could be on several of 40 pages, so being able to just change the tooltip from `ABC1` to `ABC2` in one single file is the goal

When I had the `*[ABC#]: ABC1` in the abbreviations.md it just is not detected and no tooltip on the pages is shown.

**Reply by tomvachon** [reply]
My current workaround (not ideal) was to have a material icon next to each one and put a title on that icon.  That will be a maintenance nightmare going forward

**Reply by kamilkrzyskow** [reply]
Testing the regex in the abbr processor it successfully matched the `*[ABC#]: ABC1`:
https://github.com/Python-Markdown/markdown/blob/ec8c305fb14eb081bb874c917d8b91d3c5122334/markdown/extensions/abbr.py#L103

```py
import re

pattern = r'^[*]\[(?P<abbr>[^\\]*?)\][ ]?:[ ]*\n?[ ]*(?P<title>.*)$'
RE = re.compile(pattern, re.MULTILINE)
text = '*[ABC#]: ABC1'

match = RE.search(text)
if match:
    print("Abbreviation:", match.group('abbr'))
    print("Title:", match.group('title'))
else:
    print("No match found.")
```

So the issue lies probably in the `blocks` list of strings, perhaps the `#` somehow separates blocks 🤷 
You'd need to debug it yourself or provide a complete example mkdocs project where the issue occurs, then someone might have the time to look into it.

**Reply by kamilkrzyskow** [reply]
If the backend doesn't work, frontend should handle the issues ;) So you could also have some kind of extra JavaScript that modifies the ABC1 (set via title=) to any other value.

**Reply by pawamoy** [reply]
This code is not released yet. But the regex is the same in 3.6. The issue (I think?) is that `#` gets escaped into `\#`, so it doesn't match in the rest of the document. @waylan WDYT?

**Reply by kamilkrzyskow** [reply]
👀 You say escaped with `\`, ChatGPT suggested to change the `?P<abbr>[^\\]*?` with `?P<abbr>[^\]]*?`, but as both patterns worked in my case I didn't give this much thought, but perhaps this is the way.

**Reply by Andre601** [reply]
> You say escaped with `\`, ChatGPT suggested to change the `?P<abbr>[^\\]*?` with `?P<abbr>[^\]]*?`, but as both patterns worked in my case I didn't give this much thought, but perhaps this is the way.

Well, according to Regex101 is neither even usable because `? The preceding token is not quantifiable`

And yes, I did test with the Python option.
That's honestly why I don't like nor use AI like ChatGPT. It can (and will) give responses that don't have the context necessary to reproduce stuff...

**Reply by Andre601** [reply]
Also, the main issue with both patterns is, that they trigger on basically anything, [including empty Strings](https://regex101.com/r/ND3c80/1) because the regex means "Find anything that is not `\` (or `]` with second regex) with the smallest possible match" which with 0 characters is a possability.

**Reply by kamilkrzyskow** [reply]
You only checked the insides of the abbr, there is a whole regex pattern that requires other tokens `*[]`
I agree that matching empty abbr isn't ideal, but then again people won't use empty `*[]` in Markdown

**Reply by Andre601** [reply]
> but then again people won't use empty `*[]` in Markdown

That is kinda wishful thinking. Hope for the best, but expect the worst should be the motto here...
And I tested the full regex (Which for me as a beginner seems a bit weirdly designed with all the `*` quantifiers) which seems to work too on the Python setting on regex101, so the pattern should work with abbreviations containing hashtags...

**Reply by pawamoy** [reply]
Sorry for the confusion, I meant that the matched abbreviations are escaped with `re.escape` to build a regex that contains all found abbreviations (like `\b(thing1|thing2|thing3|etc)\b`), to match them again in the body of the document, and `re.escape` actually escapes `#` into `\#` (not sure why since to me this character doesn't have a special meaning).

But it doesn't make much sense to me either, because surely if `re.escape` transforms `#` into `\#`, then `ABC\#` would match against `ABC#`?

UPDATE: yep:

```pycon
>>> re.match(re.escape("ABC#"), "ABC#")
<re.Match object; span=(0, 4), match='ABC#'>
```

**Reply by pawamoy** [reply]
Ah, however:

```pycon
>>> re.match(r"\bABC#\b", "ABC#")
>>> re.match(r"\bABC\#\b", "ABC#")
>>> 
```

(no match)

The issue seems to be the combination of a trailing `\b` after `#` or `\#`.

Any regex expert who can explain why that doesn't match? :slightly_smiling_face:

**Reply by tomvachon** [reply]
Not sure this is it, but I wonder if it being evaluated as the multiline format? https://docs.python.org/3/library/re.html#re.VERBOSE

**Comment by tomvachon** [comment]
It's at this public repo: https://github.com/virtualZOB/refs if you want to see it.  

Currently trying to match `GSH#`
Generated HTML:https://github.com/virtualZOB/refs/blob/91b820d3dc47b679e51aebfa977e4144f8df9f1a/references/LOA/Routes/ZAU/Clearance/index.html#L1823 

Markdown: https://github.com/virtualZOB/refs/blob/main/docs/references/LOA/Routes/ZAU/Clearance.md?plain=1#L9

Public page where it should be showing: https://refs.clevelandcenter.org/references/LOA/Routes/ZAU/Clearance/#clearance-routing

**Reply by pawamoy** [reply]
Thanks.

As seen [in this comment](https://github.com/mkdocs/mkdocs/discussions/3714#discussioncomment-9538451), this is a limitation/bug of the abbreviations extension. It currently does not support leading or trailing characters that are not "word" characters (because of the leading and trailing word-boundary `\b` used in the regex to match abbreviations in the body). (@waylan :wave:)

You could try to remove `#` from the reference:

```md
My text with ABC# or GSH#.

*[ABC]: ABC1
*[GSH]: GSH2
```

**Reply by tomvachon** [reply]
Thanks.  I was trying to follow but there seemed to be too many threads for me to fully follow.  I'll have to keep my current workaround and hopefully the extension is updated.  GSH and GSH# have two quite different meanings in aviation.  GSH is a VOR but GSH# is a procedure.

**Reply by waylan** [reply]
@pawamoy is correct. There must be no word boundaries within an abbreviation. Yes, this limits what an abbreviation can contain. And yes, this behavior is intentional. There is no plans to make changes to the extension to accept additional characters.

According to the [Python docs](https://docs.python.org/3/library/re.html#index-28):

> `\b` is defined as the boundary between a `\w` and a `\W` character (or vice versa), or between `\w` and the beginning or end of the string.
>...
> [`\w`] Matches Unicode word characters; this includes all Unicode alphanumeric characters (as defined by [str.isalnum()](https://docs.python.org/3/library/stdtypes.html#str.isalnum)), as well as the underscore (`_`).
> ...
> [`\W`] Matches any character which is not a word character. This is the opposite of `\w`.

What this means is that an abbreviation can only contain alphanumeric characters and the underscore.

**Reply by tomvachon** [reply]
OK, there is a chance I could maybe use _ as a place holder. thanks!
